charms/worker/k8s/src/charm.py

```diff
@@ -361,6 +362,36 @@ def _enable_functionalities(self):
 
         self.api_manager.update_cluster_config(update_request)
 
+    @on_error(
+        WaitingStatus("Ensure that the cluster configuration is up-to-date"),
+        InvalidResponseError,
+        K8sdConnectionError,
+    )
+    def _ensure_cluster_config(self):
+        """Ensure that the cluster configuration is up-to-date.
+
+        The snap will detect any changes and only perform necessary steps.
+        There is no need to track changes in the charm.
+        """
+        status.add(ops.MaintenanceStatus("Ensure cluster config"))
+        log.info("Ensure cluster-config")
+
+        update_request = UpdateClusterConfigRequest()
+
+        # TODO: Ensure other configs here as well.
+
+        if self.config.get("datastore") == "etcd":
+            etcd_config = self.etcd.get_client_credentials()
+            update_request.datastore = UserFacingDatastoreConfig(
+                type="external",
+                servers=self.etcd.get_connection_string().split(","),
+                ca_crt=etcd_config.get("client_ca", ""),
+                client_crt=etcd_config.get("client_cert", ""),
+                client_key=etcd_config.get("client_key", ""),
+            )
+
+        self.api_manager.update_cluster_config(update_request)
```

Comment:
I continue to encounter this error when the third etcd unit tries to post it's update:

```
charms.k8s.v0.k8sd_api_manager.InvalidResponseError: Error status 500
	method=PUT
	endpoint=/1.0/k8sd/cluster/config
	reason=Internal Server Error
	body={"type":"error","status":"","status_code":0,"operation":"","error_code":500,"error":"failed to reconcile network: failed to refresh network: failed to refresh network component: failed to upgrade component 'network': another operation (install/upgrade/rollback) is in progress","metadata":null}
```

Reply:
Here are the k8sd and charm logs around the same moment of this failure
https://pastebin.canonical.com/p/jD4TNbk4j4/

Reply:
api server logs around the same moment
https://pastebin.canonical.com/p/QzMpFCmQCV/

is it possible that adding extra etcd units in rapid succcession caused the API server to restart when the first datastore came online, and when we published the second datastore URL -- the config wouldn't take?

Reply:
Hey Adam,
Thanks for the review. I think the issue you were facing was before https://github.com/canonical/k8s-snap/pull/356 was merged. 

I cannot reproduce this issue and the CI also seems to be happy.